Create new page about Parachains' Protocol Overview

filippoweb3 opened the pull request:
See linked issue.

Commit:
Created new page

- added content from previous PR
- Added page to sidebar

Commit:
Added inclusion pipeline text

Commit:
Added approval process text

Commit:
Added banner with links

- to approval process and disputes

Commit:
Added summary figure

Commit:
Added new figures

- inclusion pipeline
- approval process

Commit:
added block statuses

Commit:
updated figure

- protocol overview

Commit:
Added more text to support the figures

Commit:
added new section & figure

Commit:
Added more text on asynch

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+### Fishermen
+
+They catch misbehaving validators in exchange for a bounty. Collators and validators can behave as fishermen.
```
The fishermen component has been deprecated from the Polkadot architecture https://github.com/w3f/polkadot-wiki/issues/2396

https://github.com/w3f/polkadot-wiki/issues/2504

filippoweb3 replied:
right! Will delete this part. I believe the parity's doc was written some time ago.

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+The figure below shows a representation of a parachain with collators and validators. The figure also shows the journey of a block through the three main steps, as well as the sections where the [Inclusion Pipeline](#inclusion-pipeline) and the [Approval Process](#approval-process) take place (see next Sections).
+
+![parachain-protocol-summary](../assets/parachain-protocol-summary.png)
```
Is this figure adapted from a different source? Typically, the chronological order of events is shown from left to right.

filippoweb3 replied:
I did the figures based on the description on the parity's doc. I noticed that the parachain is always shown from the right (attached to the relay chain on the right side). The idea was to keep the same location and zoom in to show the different processes (like in GIS/maps representations). That is why the figures here are from right to left. It is not clear the zoom part, I will change them a little to make the story clearer.

profradha replied:
I see! That is the architecture figure, where the position of the components does not matter. What you are describing through your figures is a pipeline (chain of events), so left to right is more intuitive I think.

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+![parachain-protocol-summary](../assets/parachain-protocol-summary.png)
+
+Negative statements will lead to a dispute, and if there are false negatives whoever will be on the wrong side (once the dispute is resolved) will be slashed. False positives can also happen, those actors responsible for it will also be slashed and whoever reported the case will get a bounty. In order to detect false positives, PoV information must be available after the block has been added to t ...
```
Whenever possible, back the statements like these "PoVs are expected to be between 1 MB and 10 MB;" with references

filippoweb3 replied:
yeah, found this in the parity's doc with no ref, maybe should I mention the document?

filippoweb3 replied:
@DrW3RK thanks for the preliminary feedback, I will get back to this PR in late January when I am back. In the meantime, I implemented the Parachain Host Implementer Guide in the current pages of the wiki. I saw there is already a page about "Availability and Validity", I think this page can be merged together with the content in this page here and create a cohesive story about the lifecycle of a parachain block.

profradha replied:
Yeah, sounds good to me!

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+We have mentioned how a relay chain block author must select the candidate and note it on the relay chain (we say the block is backed). The relay chain block author is selected by BABE, which is a forkful algorithm. This means that different block authors are chosen at the same time and they may not work on the same block parent (i.e. the representations in the previous figures are simplistic). A ...
+
+We say that the network is **asynchronous** since there will be validators who gave received a block, and other validators who did not. Thus, the network is variable and it exists in multiple states. In the figure below (*left*) the Group 1 received block C while Group 2 did not due to network asynchrony. Validators in Group 2 can build another block on top of B, called C'. Assume that afterwards ...
```
```suggestion
We say that the network is **asynchronous** since there will be validators who have received a block, and other validators who did not. Thus, the network is variable and it exists in multiple states. In the figure below (*left*) the Group 1 received block C while Group 2 did not due to network asynchrony. Validators in Group 2 can build another block on top of B, called C'. Assume that afterwards, some validators become aware of both C and C' while others remain aware of one of them (*right*). Validators in Group 3 must be aware of the network state in each head (C and C') and they may contribute to some or full extent on both. It is possible that due to network asynchrony two forks may grow in parallel for some time.
```

DrW3RK reviewed (commented):
The article reads well and the content organization is in the right direction. Great effort with the figures, much appreciated! Provided some preliminary feedback. You can keep adding content.

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Merge branch 'master' into parachain-protocol-overview-page

filippoweb3 commented:
@DrW3RK can you approve this? I will create a new PR to do the additional work of putting all info about the parachain protocol on one page, and remove the content that is spread across multiple pages (adding redirects where relevant).

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+:::
+
+The aim of the protocol is to carry a parachain's block from authoring to inclusion through a process that can be carried out repeatedly and in parallel for each parachain connected to the relay chain.
```
```suggestion
The protocol aims to carry a parachain's block from authoring to inclusion through a process that can be carried out repeatedly and in parallel for each parachain connected to the relay chain.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+### Validators
+
+They are responsible for validating proposed parachain's blocks by checking the **Proof-of-Validity** (PoV) of the blocks and ensuring the PoV remains available. They have "skin in the game", meaning they have funds bonded on-chain that can be partially or fully confiscated in case of misbehavior.
```
```suggestion
They are responsible for validating the proposed parachain's blocks by checking the **Proof-of-Validity** (PoV) of the blocks and ensuring the PoV remains available for a designated period. They have "skin in the game", meaning they have funds bonded on-chain that can be partially or fully confiscated by the network in case of misbehavior.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+### Fishermen
+
+They catch misbehaving validators in exchange for a bounty. Collators and validators can behave as fishermen.
```
```suggestion
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+They catch misbehaving validators in exchange for a bounty. Collators and validators can behave as fishermen.
+
+## Summary View
```
Keep the search engine optimization when naming sections and subsections.

```suggestion
## Parachains' Protocol Summary
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+2. Validators validate the block using the PoV, signing statements that can have positive or negative.
+3. With enough positive statements the block can be added to the relay chain.
+
+The figure below shows a representation of a parachain with collators and validators. The figure also shows the journey of a block through the three main steps, as well as the sections where the [Inclusion Pipeline](#inclusion-pipeline) and the [Approval Process](#approval-process) take place (see next Sections).
```
```suggestion
The parachains' protocol can be summarized into three main steps:

1. Collators send the parachain block (parablock) with PoV to the set for Validators assigned to the parachain.
2. Validators verify if the parablock follows the state transition rules of the parachain and sign statements that can have a positive or negative outcome.
3. With enough positive statements, the block is added to the relay chain.

The figure below shows a representation of a parachain with collators and validators. The figure also shows the journey of a block through the three main steps, as well as the sections where the [Inclusion Pipeline](#inclusion-pipeline) and the [Approval Process](#approval-process) take place (see following Sections).
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+The figure below shows a representation of a parachain with collators and validators. The figure also shows the journey of a block through the three main steps, as well as the sections where the [Inclusion Pipeline](#inclusion-pipeline) and the [Approval Process](#approval-process) take place (see next Sections).
+
+![parachain-protocol-summary](../assets/parachain-protocol-summary.png)
```
Some questions if I put myself in the shoes of a general Wiki reader

- Why is a parablock shown as a block with ?
- What is the difference between white and yellow tick marks (if they are positive and negative, then green and red colors would be a better choice).
- Can a validator sign multiple statements on the parablock or a single statement?

The gray shade can be a bit lighter.

filippoweb3 replied:
right, let me address this on Monday, I will review the figures and change the text accordingly

filippoweb3 replied:
- parablock with ? and white tick is a block that has been produced by a collator. The ? means that the block is not valid yet and the white tick is the PoV by the collator.

- the white tick is the PoV from the collator, the yellow tick are PoV from the para-validators

I can add a legend to explain the different symbols.

- as far as I know a validator can sign a single statement

Note that the network block contains multiple parablocks, I need to change this in the diagram.

I will provide an alternative diagram with lighter gray shade.

Let me know if this will answer your points :)

profradha replied:
Makes sense. Add this info as the figure's description.

filippoweb3 replied:
Alright, the information has been added in the text with the new commits.

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,107 @@
+
+![parachain-protocol-summary](../assets/parachain-protocol-summary.png)
+
+Negative statements will lead to a dispute, and if there are false negatives whoever will be on the wrong side (once the dispute is resolved) will be slashed. False positives can also happen, those actors responsible for it will also be slashed and whoever reported the case will get a bounty. In order to detect false positives, PoV information must be available after the block has been added to t ...
```
The bounty part is tied to Fishermen, which is deprecated.

```suggestion
Negative statements will lead to a dispute, and if there are false negatives, whoever will be on the wrong side (once the dispute is resolved) will be slashed. False positives can also happen; those actors responsible for it will also be slashed. To detect false positives, PoV information must be available after the block has been added to the relay chain so that validators can check the work. PoVs are typically between 1 MB and 10 MB in size and are not included in the relaychain blocks. However, as a part of the data availability scheme, they are made available on the network for a certain period so that the validators can perform the required checks.
```

DrW3RK reviewed (changes requested):
Some feedback to address. I will get to the other sections early next week.

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

Commit:
minor edit

Commit:
Modified figures

- lighter gray
- minor edits

Commit:
Added explanation for symbols in figures

Commit:
changed summary figure layout

- less details

Commit:
Minor edits to description

Commit:
Simplified figure

Removed question marks

Commit:
Minor edit approval process

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+   the availability of the candidate will be noted in the subsequent relay chain blocks.
+8. Once the relay chain state machine has enough information to consider the candidate's PoV as
+   being available, the candidate is considered to be part of the parachain and is graduated to
+   being a full parachain block.
```
```suggestion
The inclusion pipeline is the path of a parachain block (or parablock) from its creation to its
inclusion into the Relay Chain. The main checkpoints of that path are listed below.

1. Validators are assigned to parachains by the **Validator Assignment** routine.
2. A collator produces the parachain block (known as parachain candidate or candidate) along with
   PoV.
3. The collator forwards the candidate and PoV to validators assigned to the same parachain via the
   **Collator Distribution** subsystem.
4. The validators assigned to the parachain participate in the **Candidate Backing** subsystem.
   Candidates that gather enough signed validity statements are considered **"backable"** and their
   backing is the set of signed statements.
5. A relay chain block author (selected by [BABE](./learn-consensus.md#block-production-babe)) can note up to 1 backable candidate for each
   parachain to be included in the Relay Chain block alongside its backing. Once included in the
   Relay Chain the candidate is considered backable in that fork of the Relay Chain.
6. Once backable in the Relay Chain, the candidate is considered to be in "pending availability" status.
   It can only be considered a part of the parachain once it is **proven available**.
7. In the following relay chain blocks, the validators will participate in the **Availability
   Distribution** subsystem to ensure availability of the candidate. The subsequent relay chain blocks will note information regarding
      the candidate's availability.
8. Once the relay chain state machine has enough information to consider the candidate's PoV as
   being available, the candidate is considered part of the parachain and is graduated to
   being a full parachain block.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+- Candidate: A block is put forward by a collator to a validator (in this case V1). The candidate
+  block is shown as white square with one white tick mark at the side (PoV from the collator). Note
+  the candidate is not valid yet and can still fail to be included in the Relay Chain.
+- Seconded: The block is put forward by V1 to other validators (in this case V2 and V3). The
```
```suggestion
- Seconded: The block is put forward by the validator V1 to other validators (in this case V2 and V3). The
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+  it. The yellow marks show the PoV from the para-validators.
+- Backable: The block validity is attested by a majority of the validators. The backable block is
+  shown as white square with a white tick mark and three yellow tick marks on top of it. The yellow
+  marks show the PoV from the para-validators.
```
should this be "PoV statements from the para-validators"?

filippoweb3 replied:
right, specified in the text

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+
+Once the parablock is considered available and part of the parachain, is still "pending approval".
+At this stage the parablock is tentatively included in the parachain, although more confirmation is
+necessary. In fact, the validators in the parachain (i.e. the parachain validators) are sampled from
```
```suggestion
necessary. In fact, the validators assigned to the parachain (i.e. the parachain validators) are sampled from
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+allocating more validators that would ultimately reduce the throughput of the system. Because after
+being considered available a parablock will accept children blocks, failure to pass the approval
+process will invalidate the parablock and its descendants. Only the validators who backed the block
+in question will be slashed, not the validators who backed the descendants.
```
```suggestion
a validator set which is assumed to be 1/3 dishonest in the worst-case scenario. In this case, it is likely that the majority of the random para-validators sampled for a specific parachain are dishonest and can back a candidate wrongly.  To address this, the **Approval Process** allows detecting misbehavior after-the-fact without
allocating more para-validators, which would ultimately reduce the system's throughput. As a parablock can accept children blocks after being considered available, failure to pass the approval
process will invalidate the parablock as well as its descendants (children blocks). Only the validators who backed the block
in question will be slashed, not those who backed the descendants.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+1. Parablocks that have been included by the Inclusion Pipeline are pending approval for a time
+   window known as the **secondary checking window**.
+2. During the secondary checking window, validators (secondary checkers) randomly self-select to
+   perform secondary checks on the parablock.
```
```suggestion
   perform secondary checks on each of the parablock.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+enters the Approval Process. The parablock becomes accepted when it is backed, available and
+**undisputed**. Thr parablock is checked a second time by a subset of validators (V5, V6 and V7),
+and if there are no contradictory results the block is approved and gossiped to other relay chain
+validators. Note the parablock after second checks is shown as a square with white background a
```
```suggestion
validators. Note the parablock after secondary checks is shown as a square with white background a
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+**undisputed**. Thr parablock is checked a second time by a subset of validators (V5, V6 and V7),
+and if there are no contradictory results the block is approved and gossiped to other relay chain
+validators. Note the parablock after second checks is shown as a square with white background a
+yellow border enclosing a "P", and three white ticks (one for each second checker). Approved
```
```suggestion
yellow border enclosing a "P", and three white ticks (one for each secondary check). Approved
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+and if there are no contradictory results the block is approved and gossiped to other relay chain
+validators. Note the parablock after second checks is shown as a square with white background a
+yellow border enclosing a "P", and three white ticks (one for each second checker). Approved
+para-blocks are shown as yellow squares the become grey.
```
```suggestion
para-blocks are shown as yellow squares that become grey in color.
```

profradha commented on docs/learn/learn-parachains-protocol.md:
```diff
@@ -0,0 +1,184 @@
+that afterwards, some validators become aware of both C and C' while others remain aware of one of
+them (_right_). Validators in Group 3 must be aware of the network state in each head (C and C') and
+they may contribute to some or full extent on both. It is possible that due to network asynchrony
+two forks may grow in parallel for some time.
```
This section needs a conclusion. Something in the lines of "two/multiple forks may grow in parallel for some time, but eventually......"

filippoweb3 replied:
right, added

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

filippoweb3 committed:
Update docs/learn/learn-parachains-protocol.md

Co-authored-by: Radha <86818441+DrW3RK@users.noreply.github.com>

Commit:
Radha's suggestions implemented

filippoweb3 commented:
@DrW3RK I implemented your feedback and added more text addressing your suggestions. Let me know if it is ok. Next step (another PR) will be to unify the content, bringing everything into one place and deleting duplicated content.
